### PR TITLE
Combo chart sorting

### PR DESCRIPTION
**Pull request by NestorEncinas**
JIRA: TNT-456

---

Supported PR commands:

| Command                  | Description            |
| ------------------------ | ---------------------- |
| `ok to test`             | Re-run standard checks |
| `extended test`          | BackstopJS tests       |
| `extended check sonar`   | SonarQube tests        |
| `extended check cypress` | Cypress E2E tests      |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `check-extended-cypress` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6929/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8809/ : SUCCESS in 12m 25s

File: libs/sdk-ui-ext/src/internal/tests/mocks/referencePointMocks.ts [hunk 1]
```diff
@@ -1946,6 +1946,137 @@ export const multipleMetricsNoCategoriesReferencePoint: IReferencePoint = {
     },
 };
 
+export const onlyMeasuresReferencePoint: IReferencePoint = {
```
Comment (xMort) [inline libs/sdk-ui-ext/src/internal/tests/mocks/referencePointMocks.ts:1949]: There is already `oneMetricNoCategoriesReferencePoint` similar ref point. I would change it to `noMeasureInEitherBucketRefPoint` or `noMeasureInAnyBucketRefPoint` or something similar so you know just from the name that it has two measure buckets. Same applies for other reference point mocks that has two measure buckets.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6967/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8836/ : SUCCESS in 13m 41s

File: libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/comboChart/PluggableComboChart.tsx [hunk 2]
```diff
@@ -235,6 +244,76 @@ export class PluggableComboChart extends PluggableBaseChart {
...
+        buckets: IBucketOfFun[],
+        properties: IVisualizationProperties,
+    ): {
+        defaultSort: ISortConfig["currentSort"];
+        availableSorts: ISortConfig["availableSorts"];
+    } {
+        const measures = getBucketItemsByType(buckets, BucketNames.MEASURES, [METRIC]);
+        const secondaryMeasures = getBucketItemsByType(buckets, BucketNames.SECONDARY_MEASURES, [METRIC]);
+        const viewBy = getBucketItems(buckets, BucketNames.VIEW);
+        const stackMeasures = properties?.controls?.stackMeasures;
```
Comment (xMort) [inline libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/comboChart/PluggableComboChart.tsx:257]: I had an issue with this line in area chart. AD enabled it based on UI config but did not send `true` via this property unless I clicked upon it explictly. It seems that combo chart should behave the same but something is broken in AD and it is not enabled by default. Based on this method called from `getUiConfig` it should be by default enabled. https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/comboChart/PluggableComboChart.tsx#L126

I have there a fallback setup as follows: `const canSortStackTotal = properties?.controls?.stackMeasures ?? this.getUiConfig().optionalStacking.stackMeasures;` I'm not sure if it is necessary here or of there is some side effect.

File: libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/comboChart/PluggableComboChart.tsx [hunk 3]
```diff
@@ -235,6 +244,76 @@ export class PluggableComboChart extends PluggableBaseChart {
...
+        const defaultSort = viewBy.map((vb) => newAttributeSort(vb.localIdentifier, "asc"));
+
+        if (!isEmpty(viewBy) && (!isEmpty(measures) || !isEmpty(secondaryMeasures))) {
+            const mergedMeasures = [...measures, ...secondaryMeasures];
+            const areaSortEnabled = mergedMeasures.length > 1 ? true : false;
```
Comment (xMort) [inline libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/comboChart/PluggableComboChart.tsx:262]: You can just do `const areaSortEnabled = mergedMeasures.length > 1;` ;)

File: libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/comboChart/PluggableComboChart.tsx [hunk 4]
```diff
@@ -235,6 +244,76 @@ export class PluggableComboChart extends PluggableBaseChart {
...
+            return {
+                defaultSort,
+                availableSorts: [
+                    {
+                        itemId: localIdRef(viewBy[0].localIdentifier),
+                        attributeSort: {
+                            areaSortEnabled,
+                            normalSortEnabled: true,
+                        },
+                        metricSorts: [
```
Comment (xMort) [inline libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/comboChart/PluggableComboChart.tsx:288]: Why the array copying? Isn't `metricSorts: mergedMeasures.map((m) => newMeasureSortSuggestion(m.localIdentifier))` enough? The `map` function already creates a new array.

File: libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/comboChart/PluggableComboChart.tsx [hunk 5]
```diff
@@ -235,6 +244,76 @@ export class PluggableComboChart extends PluggableBaseChart {
...
+            availableSorts: [],
+        };
+    }
+
+    public getSortConfig(referencePoint: IReferencePoint): Promise<ISortConfig> {
+        const { buckets, properties } = referencePoint;
+        const measures = getBucketItemsByType(buckets, BucketNames.MEASURES, [METRIC]);
+        const secondaryMeasures = getBucketItemsByType(buckets, BucketNames.SECONDARY_MEASURES, [METRIC]);
+        const viewBy = getBucketItems(buckets, BucketNames.VIEW);
+        const disabled = viewBy.length < 1 || (measures.length < 1 && secondaryMeasures.length < 1);
```
Comment (xMort) [inline libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/comboChart/PluggableComboChart.tsx:307]: Please extend this condition with `... || availableSorts.length === 0;`.
This requires to switch line 307 with line 308. I would want returned sort config to be in correct state and remove condition from AD that checks `availableSorts` length and disables button if it is empty.

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/1373/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6973/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/1373/ : UNSTABLE in 22m 50s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6976/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8841/ : SUCCESS in 12m 24s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6988/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8850/ : SUCCESS in 10m 58s

**Review by xMort — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR2280

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/2034/ : SUCCESS in 1m 06s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/2317/ : SUCCESS in 19s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1127/ : SUCCESS in 15m 54s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8869/ : SUCCESS in 11m 04s

**Pull request merged**